Added a footer to informational pages on the site

garbanzos opened the pull request:
Fixes #1615. 

![capture](https://cloud.githubusercontent.com/assets/12397473/13963414/35ca2af8-f0a0-11e5-91fa-f10d3debee33.PNG)

seanlip commented:
Is it possible to align the left margin of the footer with the left margin of the text in the main card (such that this is the case on various screen sizes)? Not sure if that's tricky, but it may look better...

Also, does this behave nicely on mobile? If not, it might be worth figuring out what's best to do there, and implementing that as well.

seanlip commented:
I actually meant something slightly different (aligning it with the text in the card, rather than the left edge of the card) but I think what you've done is also fine in the interim (especially since the About page is currently undergoing a redesign anyway). So yep, let's go with it!

And yep, I think the main remaining issue is the mobile one. I also noticed one more thing: we don't really have "Featured Topics" under "LEARN", yet, so can we just have a single item "Browse Gallery"? We can easily modify it later.

Thanks!

seanlip commented on core/templates/dev/head/footer.html:
```diff
@@ -0,0 +1,59 @@
+<!DOCTYPE html>
```
Don't need this, and don't need 'html' tag below. This is an HTML fragment which will be inserted in the page.

seanlip commented on core/templates/dev/head/footer.html:
```diff
@@ -0,0 +1,59 @@
+<!DOCTYPE html>
+<html>
+  <style>
+    .footer{
```
style nit: follow oppia.css -- single space before {, indent subsequent lines by 2 spaces, alphabetize CSS attributes within a rule.

seanlip commented on core/templates/dev/head/footer.html:
```diff
@@ -0,0 +1,59 @@
+
+  <footer class="footer">
+    <div class="footer-column">
+      <span>ABOUT OPPIA</span>
```
Can you use a header element for this (more semantic for accessibility reasons) and use a "text-transform: capitalize" CSS rule instead?

seanlip commented on core/templates/dev/head/footer.html:
```diff
@@ -0,0 +1,59 @@
+  <footer class="footer">
+    <div class="footer-column">
+      <span>ABOUT OPPIA</span>
+      <a href="/about">About</a><br>
```
Prefer not to use <br>. This is really a list, so can you use ul/li and add CSS to remove the bullet points?

seanlip commented on core/templates/dev/head/profile/signup.html:
```diff
@@ -156,6 +156,7 @@ <h2 class="oppia-signup-page-title">Complete Your Registration</h2>
       <button class="btn btn-default" ng-click="close()">Close</button>
     </div>
   </script>
+  {% include 'footer.html' %}
```
Actually let's drop the inclusion on this page, since the page is quite action-oriented.

garbanzos commented:
I made changes according to your suggestions. 
The footer also displays correctly on mobiles now.

garbanzos commented:
@seanlip This PR is ready for review again!

seanlip commented:
Sorry for the delay! I was away.

A few UI issues:
- On large screens, the footer text is not centered; it looks quite odd. (You can simulate this by zooming out.)
- It would be nice if the column headers were bold, to differentiate them more from the links below them.
- Can we change "Browse" to "Browse the Gallery"?
- The footer jumps up a bit too much in the profile page (if the screen height is large). I suggest keeping the top margin of the footer where it currently is, but changing the grey space below the footer to match the dark green of the footer, so that the page doesn't have a sort of "waistband".

seanlip commented:
Some more thoughts:
- Maybe add a "margin-bottom: 10px;" to the column headers in the footer as well, to separate them from the links below.
- Re the fourth point in the previous comment, I think a reasonable alternative behavior is to keep the footer stuck to the bottom of the page (so, it basically has a maximum height). Feel free to implement whichever suggestion is easier (or whichever one you think looks better)!

seanlip commented:
Hi @garbanzos -- how is this going?

garbanzos commented:
@seanlip Sorry I was busy doing some school work. 
I think I can make the changes by this Wednesday (6 April).

seanlip commented:
Sounds good, thanks @garbanzos!

garbanzos commented:
@seanlip Ready for review!
Here are the changes I made: 
1. The footer is now centered on large screens too. 
2. The style of headers are changed - now bold and has a bottom margin of 10px 
3. Changed "Browse" to "Browser The Gallery"
4. The footer is stuck to the bottom of the profile page. 

For point 4, I am not very sure about the solution. I tried numerous methods but they didn't work. :sweat: 
It somehow footer got stuck to the bottom when I jinja include the footer in the footer_js block in the profile page. :astonished:  
Do you think it is ok if I do that?

seanlip commented:
Thanks, @garbanzos! 

Hmm, I think it's a bit odd to include it in footer_js. However, that does suggest a solution. Why not add a new Jinja block called 'footer' in base.html? Then, wherever a footer is required, use that block.

garbanzos commented:
@seanlip I tried adding `{% block footer %}{% endblock footer %}`
after `{% block footer_js %}{% endblock footer_js %}` in base.html and jinja included footer.html in the pages:  
```
{% block footer %} 
  {% include 'footer.html' %} 
{% endblock footer%}
```
Not sure if I'm doing it correctly but this works for the profile page.  However, jinja including the footer.html in the footer block doesn't seems to work out for the rest of the pages (about page etc.).  

Should I do this only for the profile page and continue to jinja include the footer in the content block for the rest of the pages?

seanlip commented:
Hmm -- you *could*, but that's a hack. Instead, it would be better to figure out what the difference is between the profile page and the other pages that's causing the behaviour to be different. Would you be able to do that?

garbanzos commented:
@seanlip I'm really sorry but I'm still not sure why this happens after some investigation. 

Furthermore, it seems the method that I used to keep the footer down for the profile page is also wrong. (It was a coincidence that it worked for the profile page) The method did not work on the preference page which also had the same "waistband" problem as the profile page. 

Could you merge this PR after I reset my last commit? And make an issue for someone else to solve the "waistband" problem?

I'm really sorry!

seanlip commented:
OK, thanks @garbanzos! One note, though: if it doesn't work out, could you explain the various methods you tried (and what went wrong with each), and why it was a coincidence that it worked for the profile page? This will help future developers not reinvent the wheel, and understand better what is going on. (Of course, it's best if you can tie it up and fix it, but if you can't, then at least we know what not to try :-) )

garbanzos commented:
@seanlip Ready for review again! 

Here are the changes I made: 
- The footer is now stuck to the bottom of all the relevant pages  
- Added a footer block in base.html 
- footer.html is now included in the footer block of relevant pages 
- Moved all the css of the footer into oppia.css

garbanzos commented:
When I was testing my code I also realise that the text field for Preferred Languages in the Preferences page don't seem to be displaying well on mobile phones.  Should I submit an issue for it? 

![image](https://cloud.githubusercontent.com/assets/12397473/14352046/1bb628da-fd05-11e5-81cc-a01e96764577.png)

seanlip commented:
Thanks, @garbanzos, that's great! I'll look at the changes later, but in the meantime, yep, please go ahead and file a new issue for the UI problem you pointed out. (Feel free to do this anytime you spot a problem, too.) Thanks!

seanlip commented on core/templates/dev/head/error/error.html:
```diff
@@ -54,4 +54,5 @@
     </div>
 
   </div>
+  {% include 'footer.html' %}
```
Just to check, why is this not in the footer block?

garbanzos replied:
Oops I missed out the error page. I will change this!

seanlip commented:
This looks _excellent_, @garbanzos -- thanks very much! I had only one question, but otherwise it looks good to me and can be merged in.

garbanzos commented:
@seanlip Ready for review again! 
I included footer.html in the footer block instead of the content block of the error page.

seanlip commented:
Looks great, thanks! I'll merge when Travis checks pass.

BTW, please don't forget to file the issue you noticed at https://github.com/oppia/oppia/pull/1632#issuecomment-206891757 -- we don't want it to slip through the cracks. Thanks!

garbanzos commented:
Ok I just filed the issue. :)

seanlip commented:
Thanks, @garbanzos! Merging :-)
